src/shims/thread.rs

```diff
@@ -0,0 +1,138 @@
+use crate::*;
+use rustc_index::vec::Idx;
+use rustc_target::abi::LayoutOf;
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn pthread_create(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        _attr: OpTy<'tcx, Tag>,
+        start_routine: OpTy<'tcx, Tag>,
+        arg: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.tcx.sess.warn(
+            "thread support is experimental. \
+             For example, Miri does not detect data races yet.",
+        );
+
+        let new_thread_id = this.create_thread()?;
+        let old_thread_id = this.set_active_thread(new_thread_id)?;
+
+        let thread_info_place = this.deref_operand(thread)?;
+        let thread_info_type = thread.layout.ty
+            .builtin_deref(true)
+            .ok_or_else(|| err_ub_format!(
+                "wrong signature used for `pthread_create`: first argument must be a raw pointer."
+            ))?
+            .ty;
+        let thread_info_layout = this.layout_of(thread_info_type)?;
```

Comment:
This seems unnecessarily roundabout; `thread_info_place.layout` should already be the layout you are looking for.